[#11] 사용자의 언어레벨을 변경할 수 있는 기능구현

msugo1 opened the pull request:
## Controller 계층 관련

### LanguageController (new)

- 사용자가 등록한 언어 레벨 변경 요청을 받을 `modifyLanguageLevel` 메소드 추가

## Service 계층 관련

### LanguageService (new)

- 언어 레벨 변경 요청에 대한 로직을 처리할 `modifyLevel` 메소드 추가

1. 등록된 언어 status 가 `모국어(NATIVE)` 인 언어들의 경우에는 레벨
   변경이 불가능 (예외)

2. 필요시 여러 개의 언어레벨 변경 요청을 한번에 받아 처리해야
   `@Transactional` annotation 추가

## Mapper 계층 관련

### LanguageMapper

- 서비스 계층에서 처리한 언어레벨 변경에 대한 요청을 DB에 반영하기 위한
  `updateLevel` 메소드 작성

### LangaugeMapper.xml

- 위의 메소드를 처리하기 위한 Query 추가

## application.properties

- Mybatis가 한 번에 여러 요청을 처리할 수 있도록 `allowMultipleQueries` 속성값 true
  추가

## 테스트 관련

- 위에서 작성한 로직을 통합 테스트 할 클래스 추가

- 포스트맨을 활용해 구현한 로직 테스트

Commit:
[#11] 사용자의 언어레벨을 변경할 수 있는 기능구현

Controller 계층 관련

* LanguageController (new)

- 사용자가 등록한 언어 레벨 변경 요청을 받을 `modifyLanguageLevel` 메소드 추가

Service 계층 관련

* LanguageService (new)

- 언어 레벨 변경 요청에 대한 로직을 처리할 `modifyLevel` 메소드 추가

1. 등록된 언어 status 가 `모국어(NATIVE)` 인 언어들의 경우에는 레벨
   변경이 불가능 (예외)

2. 필요시 여러 개의 언어레벨 변경 요청을 한번에 받아 처리해야
   `@Transactional` annotation 추가

Mapper 계층 관련

* LanguageMapper

- 서비스 계층에서 처리한 언어레벨 변경에 대한 요청을 DB에 반영하기 위한
  `updateLevel` 메소드 작성

* LangaugeMapper.xml

- 위의 메소드를 처리하기 위한 Query 추가

application.properties

* Mybatis가 한 번에 여러 요청을 처리할 수 있도록 `allowMultipleQueries` 속성값 true
  추가

테스트 관련

- 위에서 작성한 로직을 통합 테스트 할 클래스 추가

- 포스트맨을 활용해 구현한 로직 테스트

f-lab-dev commented on src/main/resources/mappers/LanguageMapper.xml:
```diff
@@ -27,4 +27,13 @@
     </select>
+
+    <update id="updateLevel" parameterType="map">
+        <foreach collection="langNewLevel" item="lang" separator=";">
```
이렇게하면 UPDATE 구문이 여러개가 실행될텐데 구문 한개로 같이 실행시킬 수 있는 방법은 없을까요?

msugo1 replied:
조언주신대로 `update ~ case when ~ then` 을 이용해서 하나로 실행시키도록 변경하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/LanguageController.java:
```diff
@@ -18,4 +18,10 @@ public void addLanguages(@CurrentUser String userId, @RequestBody LanguageDataWr
     }
+
+    @PutMapping
+    public void modifyLanguageLevel(@CurrentUser String userId, @RequestBody LanguageDataWrapper languageDataWrapper) {
```
여기서 `LanguageDataWrapper`는 Request용 클래스이니까 `DataWrapper` 보다는 `LanguageModifyRequest` 이런거로 해주는게 낫지 않을까요?

msugo1 replied:
`LanguageDataWrapper` 클래스가 `add/modify` 시 모두 사용되는 점에 착안해서 `LanguageUpsertRequest`로 변경하였습니다.

Commit:
[#11] 피드백 반영

- `LanguageDataWrapper -> LanguageUpsertRequest` 로 클래스 명칭 변경
: 사용자의 언어 추가/변경에 대한 요청에 사용되는 클래스로 그 용도가 더욱
명확해지도록

- `updateLevels` 쿼리 변경
: 기존에 개별로 실행되던 쿼리들을 `UPDATE ~ CASE WHEN ~ THEN`을 이용해
한 번에 실행되도록 변경

- 변수 이름 변경

f-lab-dev reviewed (approved):
👍
